### ISSUE TITLE
Linter idea: z-index max value

### ISSUE THREAD

**Issue by mattdell** [brigade/scss-lint#460]
It would be great to be able to restrict the z-index value across a project.

For example, enforcing a maximum value of 10 instead of someone starting with 999, the next person coming along and typing 9999, and going all the way up to 999999999999. The result should be z-index values between 1 and 10. I can't think of many instances where a site should need more than 10 layers of z-index.

Thoughts?

**Comment by Anahkiasen** [comment]
:+1:

**Comment by trotzig** [comment]
Alternatively, we could also write a linter that behaves similar to the [ColorVariable linter](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#colorvariable), where only variables are valid for z-index. Then you can have a global configuration for all z-indexes and just reference them when you need them. This is what we do at Brigade. Here's our global configuration:

```scss
$z-index-menu: 100;
$z-index-navigation-main: 200;
$z-index-floating-action-button: 300;
$z-index-modal-wrapper: 400;
$z-index-toast: 500;
$z-index-n-progress: 600;
```

I'm not actually sure why we leave a one-hundred gap in between. Since we have full control over the values, we might as well just use 1-6.

**Comment by mattdell** [comment]
That's also a good suggestion, @trotzig. It's funny you mention that solution because I just wrote everything to go into a configuration file for my current project.

**Comment by Anahkiasen** [comment]
Well currently the variables thing is easy to do by adding `z-index` to the VariablesForProperty linter

**Comment by trotzig** [comment]
Ah, I forgot about that linter! Great that you mentioned it, I just spent a bit of time making sure that all our z-indexes were in fact variables (a few instances didn't use variables in our codebase). I'll make sure to add that to our project's scss-lint config.

**Comment by mattdell** [comment]
I've added that to our scss-lint config. Works great!

**Comment by trotzig** [comment]
I'm closing this issue now as it seems like we have something good enough in place already. Feel free to reopen if you think we need a linter more closely targeted at z-indexes.

**Comment by Anahkiasen** [comment]
Well no, the original issue was about preventing z-indexes from being too random

**Comment by trotzig** [comment]
Alright, cool. I'm happy to keep this open until we have something else. The VariableForProperty linter seemed enough to me. The upside of having that linter too is that you have to semantically name your z-index values. The names make it easier to reason about what the value means. Though I have to admit that it's hard to come up with good names.

**Comment by sds** [comment]
@Anahkiasen Can you elaborate on why using `VariableForProperty` doesn't solve that problem?

I would suggest using variables is the _best_ way to ensure your z-indexes are "less random." You can define semantically named variables describing different "layers" of depths so you can easily deduce from the variable name which layer an element resides.

If using `VariableForProperty` does not allow you to enforce a specific set of z-indexes, what kind of system would you consider? I would think that having a linter which enforces a specific range of z-index values is pretty arbitrary and doesn't give you any semantic benefit from a readability perspective. Someone new to your project doesn't know the consequence of `z-index` 100 or 1000 relative to other z-indexes, but variables for the indexes would.

Interested to hear your thoughts.

**Comment by Anahkiasen** [comment]
Just basically a linter that takes the lowest and highest z-index and asserts that the range isn't greater than a configured X number, default being 10 per example or something. While VariableForProperty solves half of the problem, it doesn't prevent a bad developer from spamming nonsensical z-indexes like 99999999.

**Comment by sds** [comment]
@Anahkiasen: I believe `VariableForProperty` does solve the problem. By having a predefined set of z-indexes, your developers always choose from that predefined set. If they try to create a non-sensical number, `VariableForProperty` prevents them from doing this since it enforces you use a variable.

If your issue is that your developers would add to this pre-defined list of z-indexes, that's something that can easily be corrected. As long as you're using variables, the actual value of the z-index doesn't really matter unless you're integrating with third-party CSS code.

If your issue is that developers would create variables inline to skirt around the issue, it sounds like you have a process problem. :)

Adding a linter to enforce a specific integer range for z-indexes doesn't seem to solve the actual problem, unless there's a particular situation you're thinking of that I'm not considering here.

**Comment by sds** [comment]
Closing since I'm not seeing convincing arguments for the need for such a linter, especially given the high-level issue is solvable using `VariableForProperty`, as discussed above.

Feel free to reopen if you have strong feelings otherwise, but at that point a pull request would be preferred.
